File: tests/gsl.ext/enums.py
#!/usr/bin/env python3
# -*- Python -*-
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2026 all rights reserved


"""
Exercise the bound flag enumerations directly, including that the blas and eigen bindings take
them by type, so an integer or a bad value is rejected rather than silently accepted
"""


def test():
    # the bindings
    from gsl import libgsl

    # each enumeration publishes the expected members
    assert set(libgsl.Transpose.__members__) == {"noTranspose", "transpose", "conjugateTranspose"}
    assert set(libgsl.Triangle.__members__) == {"upper", "lower"}
    assert set(libgsl.Diagonal.__members__) == {"unit", "nonUnit"}
    assert set(libgsl.Side.__members__) == {"left", "right"}
    assert set(libgsl.EigenOrder.__members__) == {
        "valueAscending",
        "valueDescending",
        "magnitudeAscending",
        "magnitudeDescending",
    }

    # the values carry gsl's own cblas numbers, so they pass straight into the library
    assert int(libgsl.Transpose.noTranspose) == 111
    assert int(libgsl.Triangle.upper) == 121
    assert int(libgsl.Diagonal.unit) == 132
    assert int(libgsl.Side.left) == 141

    # a blas call that takes a flag accepts the enum
    A = libgsl.Matrix(shape=(2, 2)).identity()
    x = libgsl.Vector(shape=2).fill(1.0)
    y = libgsl.Vector(shape=2).zero()
    libgsl.blas_dgemv(libgsl.Transpose.noTranspose, 1.0, A, x, 0.0, y)
    assert y.get(0) == 1.0

    # but a raw integer in place of the enum is rejected
    try:
        # pass an int where a {Transpose} is expected
        libgsl.blas_dgemv(0, 1.0, A, x, 0.0, y)
    # which is what should happen
    except TypeError:
        pass
    # and if it doesn't
    else:
        # the type is not being enforced
        assert False, "a blas flag should be an enum, not an int"

    # all done
    return


# main
if __name__ == "__main__":
    # do...
    test()


# end of file

File: extensions/gsl/enums.cc
// -*- C++ -*-
// -*- coding: utf-8 -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2026 all rights reserved


// external dependencies
#include "external.h"
// namespace setup
#include "forward.h"
// the blas flag types and the eigenvalue ordering
#include <gsl/gsl_blas.h>
#include <gsl/gsl_eigen.h>


// add the enumerations to the module
//
// these carry gsl's own values, so they pass straight into the library with no translation; the
// blas and eigen bindings take them by these types, which makes an invalid flag impossible to
// spell rather than a runtime complaint
void
gsl::py::enums(py::module & m)
{
    // whether a blas operand is used plain, transposed, or conjugate transposed
    py::enum_<CBLAS_TRANSPOSE_t>(m, "Transpose", "how a blas operand is used")
        .value("noTranspose", CblasNoTrans, "use the matrix as it is")
        .value("transpose", CblasTrans, "use the transpose of the matrix")
        .value("conjugateTranspose", CblasConjTrans, "use the conjugate transpose of the matrix");

    // which triangle of a symmetric or triangular matrix a blas call reads
    py::enum_<CBLAS_UPLO_t>(m, "Triangle", "which triangle of a matrix a blas call reads")
        .value("upper", CblasUpper, "the upper triangle and the diagonal")
        .value("lower", CblasLower, "the lower triangle and the diagonal");

    // whether a triangular matrix has an implicit unit diagonal
    py::enum_<CBLAS_DIAG_t>(m, "Diagonal", "whether a triangular matrix has a unit diagonal")
        .value("unit", CblasUnit, "the diagonal entries are taken to be one")
        .value("nonUnit", CblasNonUnit, "the diagonal entries are as stored");

    // which side of a product a matrix sits on
    py::enum_<CBLAS_SIDE_t>(m, "Side", "which side of a product a matrix sits on")
        .value("left", CblasLeft, "the matrix multiplies from the left")
        .value("right", CblasRight, "the matrix multiplies from the right");

    // the order in which an eigensolver returns its eigenvalues and eigenvectors
    py::enum_<gsl_eigen_sort_t>(m, "EigenOrder", "the ordering of an eigensystem")
        .value("valueAscending", GSL_EIGEN_SORT_VAL_ASC, "by value, smallest first")
        .value("valueDescending", GSL_EIGEN_SORT_VAL_DESC, "by value, largest first")
        .value("magnitudeAscending", GSL_EIGEN_SORT_ABS_ASC, "by magnitude, smallest first")
        .value("magnitudeDescending", GSL_EIGEN_SORT_ABS_DESC, "by magnitude, largest first");

    // all done
    return;
}


// end of file
